framework/finalizer.go

```diff
@@ -0,0 +1,99 @@
+package framework
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	finalizerName = "operatorkit.giantswarm.io"
+)
+
+type patchSpec struct {
+	Op    string   `json:"op"`
+	Path  string   `json:"path"`
+	Value []string `json:"value"`
+}
+
+func (f *Framework) addFinalizer(obj interface{}) (bool, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return false, err
+	}
+	if accessor.GetDeletionTimestamp() != nil {
+		return true, nil // object has been marked for deletion, we should ignore it.
+	}
+	if containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		return false, nil // object already has the finalizer.
+	}
+
+	c, err := rest.RESTClientFor(&rest.Config{})
+	if err != nil {
+		return false, err
+	}
+	patch := fmt.Sprintf(`{"op":"add","value":%q,"path":"/metadata/finalizers/1"}`, finalizerName)
+	res := c.Patch(types.JSONPatchType).
+		AbsPath(accessor.GetSelfLink()).
+		Body(patch).
+		Do()
+	if res.Error() != nil {
+		return false, res.Error()
+	}
+
+	return true, nil
+}
+
+func (f *Framework) removeFinalizer(obj interface{}) error {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+	if !containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		return nil // resource has no finalizer, probably migration.
```

Comment:
We could remove this after we have done migration of all operators and return an error instead?

Reply:
I would just keep it and add a warning log. This can potentially always happen. Also for users using operatorkit outside Giant Swarm.